src/websockets/legacy/protocol.py

```diff
@@ -132,6 +132,19 @@ def __init__(
         self.read_limit = read_limit
         self.write_limit = write_limit
 
+        # Unique identifier. For logs.
+        self.id = uuid.uuid4()
+
+        # Logger or LoggerAdapter for this connection.
+        if logger is None:
+            logger = logging.getLogger("websockets.protocol")
+        # https://github.com/python/typeshed/issues/5561
+        logger = cast(logging.Logger, logger)
+        self.logger = logging.LoggerAdapter(logger, {"websocket": self})
```

Comment:
this breaks uvicorn because `self.logger.level` no longer exists: https://github.com/encode/uvicorn/blob/e881ab06d590adbf3aa5c8333be2306abb50d077/uvicorn/protocols/websockets/websockets_impl.py#L79-L80

Reply:
As a general good practice, I would recommend you use public APIs rather than private APIs.

The `level` attribute of `logging.Logger` objects is a private API: it isn't documented here: https://docs.python.org/3/library/logging.html#logging.Logger

The `getEffectiveLevel()` is the public API for doing what you want and it works identically for `logging.Logger` and `logging.LoggerAdapter` objects:


```python
>>> import logging
>>> logger = logging.getLogger()
>>> adapter = logging.LoggerAdapter(logger)
>>> logger.level  # PRIVATE API! DON'T DO THIS!
30
>>> logger.getEffectiveLevel()
30
>>> adapter.level
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'LoggerAdapter' object has no attribute 'level'
>>> adapter.getEffectiveLevel()
30
```